src/cluster_legacy.c

```diff
@@ -5233,6 +5251,7 @@ int clusterAddSlot(clusterNode *n, int slot) {
     if (server.cluster->slots[slot]) return C_ERR;
     clusterNodeSetSlotBit(n,slot);
     server.cluster->slots[slot] = n;
+    clearCachedClusterSlotsResponse();
```

Comment:
rather than spreading `clearCachedClusterSlotsResponse`, I wonder if it suffices to clear the cache on any config save operation, i.e. in `clusterSaveConfigOrDie`? I can't think of a case where we would not save nodes.conf on topology changes.

Reply:
I had considered this at one point, but it also would cause unnecessary invalidation for cases, since we often unconditionally update the config. It looks really close and seems like a much more maintainable backstop than all these scattered hooks though.

Reply:
I am less concerned about the premature cache invalidation unless there is a significant difference. I do think it is not easy to track the many invalidation calls and reason about them. The mental model is a lot cleaner if we can tie it to the config save path, assuming the difference is trivial.

Reply:
Will have have to again capture the cases which are not covered by `clusterSaveConfigOrDie`. But it will surely as said by @madolson, it would also cause unnecessary invalidation for cases where it would not be necessary otherwise.

Reply:
As a side note, we should investigate unnecessary config save calls. This not only impacts performance (stalling the main thread due to synchronous I/O and now premature cache invalidation) but also increases the risk of disk I/O failures taking down the Valkey server.

Reply:
FWIW, AWS doesn't use the nodes.conf file persistence. We don't ever reboot nodes, we just replace them and restore the state manually. This is to avoid all the issues you mentioned with stalling I/O.

Reply:
Good to know but that is a vendor specific change that trades complexity for reliability/performance. There is still value in getting more disciplined in the Valkey codebase.